### PR TITLE
propagate the error from parsing enum variant to the parser and emit out

### LINKED ISSUE #103869: Error message when defining enum using struct syntax
Although I know Enums take to form of:
```rust
enum VecOrMap{
    Vec(Vec<usize>),
    Map(HashMap<String,usize>)
}
```
I sometimes find myself accidentally defining them as I would a Struct ([playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=8d9f591838885f65ebf69ff075a3eae9)):
```rust
enum VecOrMap{
    vec: Vec<usize>,
    map: HashMap<String,usize>
}
```

The current output is:

```
   Compiling playground v0.0.1 (/playground)
error: expected one of `(`, `,`, `=`, `{`, or `}`, found `:`
 --> src/lib.rs:2:8
  |
2 |     vec: Vec<usize>,
  |        ^ expected one of `(`, `,`, `=`, `{`, or `}`

error: expected item, found `:`
 --> src/lib.rs:2:8
  |
2 |     vec: Vec<usize>,
  |        ^ expected item

error: could not compile `playground` due to 2 previous errors
```

This always takes more than a few seconds to parse and then realize my mistake. 
Is it possible to change this to something like:

```
It looks like you are trying to define an Enum using Struct syntax, change `vec: Vec<usize>` to `Vec(Vec<usize>)`
```

### PR DESCRIPTION
**Pull request by lyming2007**
While parsing enum variant, the error message always disappear
Because the error message that emit out is from main error of parser
The information of enum variant disappears while parsing enum variant with error
We only check the syntax of expecting token, i.e, in case https://github.com/rust-lang/rust/issues/103869
It will error it without telling the message that this error is from pasring enum variant.
Propagate the sub-error from parsing enum variant to the main error of parser by chaining it with map_err
Check the sub-error before emitting the main error of parser and attach it.
Fix https://github.com/rust-lang/rust/issues/103869

### CODE REVIEW

File: compiler/rustc_parse/src/parser/item.rs [hunk 1]
```diff
@@ -1414,7 +1414,11 @@ impl<'a> Parser<'a> {
 
                 Ok((Some(vr), TrailingToken::MaybeComma))
             },
-        )
+        );
+        if let Err(ref mut err) = result_tt {
+            err.help("enum variants can be `Variant` or `Variant = <integer>` or `Variant(Type)` or `Variant { field: Type }`");
+        }
+        return result_tt;
```
Comment (fbstj) [inline compiler/rustc_parse/src/parser/item.rs:1421]:
```suggestion
        result_tt
```
though it's possible the `collect_tokens_trailing_token` can be chained with a `map_err` or an `inspect_err` call?

Comment (lyming2007) [reply compiler/rustc_parse/src/parser/item.rs]: yes collect_tokens_trailing_token can be chained with a map_err

File: compiler/rustc_parse/src/parser/mod.rs [hunk 2]
```diff
@@ -943,6 +943,10 @@ impl<'a> Parser<'a> {
                                 Err(e) => {
                                     // Parsing failed, therefore it must be something more serious
                                     // than just a missing separator.
+                                    for xx in &e.children {
+                                        // propagate the help message from sub error 'e' to main error 'expect_ee;
```
Comment (Rageking8) [inline compiler/rustc_parse/src/parser/mod.rs:947]:
```suggestion
                                        // propagate the help message from sub error 'e' to main error 'expect_err'
```
Is this comment supposed to be like that?

Comment (lyming2007) [reply compiler/rustc_parse/src/parser/mod.rs]: fix was pushed. thanks for the correction

File: src/test/ui/structs/struct-fn-in-definition.stderr [hunk 3]
```diff
@@ -33,6 +33,7 @@ LL |     fn foo() {}
    |
    = help: unlike in C++, Java, and C#, functions are declared in `impl` blocks
    = help: see https://doc.rust-lang.org/book/ch05-03-method-syntax.html for more information
+   = help: enum variants can be `Variant` or `Variant = <integer>` or `Variant(Type)` or `Variant { field: Type }`
```
Comment (eholk) [inline src/test/ui/structs/struct-fn-in-definition.stderr:36]: The extra help message seems a little unexpected to me. It's technically true, but I don't think it's really the problem the user is running into here. I'm not sure it's worth the extra logic to differentiate this case though?

**Comment by eholk** [comment]
Thanks for the contribution! Overall, this looks pretty good. I posted a couple of questions about minor tweaks to the hint message, but I think this is a big improvement.

One suggestion I'd add for future PRs. In the commit description, rather than just saying "fix https://github.com/rust-lang/rust/issues/103869," it'd be helpful if you could include a brief summary of the bug you're fixing (just a sentence or two and maybe a short code example if it makes sense) and how your PR fixes it (this isn't always going to be that interesting, but especially if there's something subtle about the fix this can be really helpful). As a reviewer that helps a lot because it saves me a step in having to pull up the issue and read the description first.

Thanks again! I'll give you a little time in case you want to respond to any of my review comments, but I don't think I'd consider any of them blockers.

File: src/test/ui/structs/struct-fn-in-definition.stderr (hunk 3 above)
Comment (lyming2007) [reply src/test/ui/structs/struct-fn-in-definition.stderr]: changed the suggestion message to:
`enum variants can be `Variant`, `Variant = <integer>`, `Variant(Type, ..., TypeN)` or `Variant { fields: Types }``

**Comment by lyming2007** [comment]
> Thanks for the contribution! Overall, this looks pretty good. I posted a couple of questions about minor tweaks to the hint message, but I think this is a big improvement.
> 
> One suggestion I'd add for future PRs. In the commit description, rather than just saying "fix #103869," it'd be helpful if you could include a brief summary of the bug you're fixing (just a sentence or two and maybe a short code example if it makes sense) and how your PR fixes it (this isn't always going to be that interesting, but especially if there's something subtle about the fix this can be really helpful). As a reviewer that helps a lot because it saves me a step in having to pull up the issue and read the description first.
> 
> Thanks again! I'll give you a little time in case you want to respond to any of my review comments, but I don't think I'd consider any of them blockers.
Thank you Eric for the feedback! I've updated the commit description. Here is it:
_While parsing enum variant, the error message always disappear
Because the error message that emit out is from main error of parser
The information of enum variant disappears while parsing enum variant with error
We only check the syntax of expecting token, i.e, in case #103869
It will error it without telling the message that this error is from pasring enum variant.
Propagate the sub-error from parsing enum variant to the main error of parser by chaining it with map_err
Check the sub-error before emitting the main error of parser and attach it.
Fix #103869_

**Comment by eholk** [comment]
Thanks, looks good!

@bors r+

**Pull request merged**